docs/pluggable-discovery-specification.md

```diff
@@ -0,0 +1,368 @@
+new port is detected or removed respectively. If the discovery goes into "events" mode successfully the response to this
+command is:
+
+```JSON
+{
+  "eventType": "start_sync",
+  "message": "OK"
+}
+```
+
+After this message the discoery will send `add` and `remove` event asyncronoushly (more on that later). If an error
+occurs and the discovery can't go in "events" mode the error must be reported as:
+
+```JSON
+{
+  "eventType": "start_sync",
+  "error": true,
+  "message": "Resource busy"
+}
+```
+
+The `error` field must be set to `true` and the `message` field should contain a description of the error.
+
+Once in "event" mode, the discovery is allowed to send `add` and `remove` messages asynchronously in realtime, this
+means that the client must be able to handle these incoming messages at any moment.
+
+The `add` event looks like the following:
+
+```JSON
+{
+  "eventType": "add",
+  "port": {
+    "address": "/dev/ttyACM0",
+    "label": "ttyACM0",
+    "properties": {
+      "pid": "0x804e",
+      "vid": "0x2341",
+      "serialNumber": "EBEABFD6514D32364E202020FF10181E",
+      "name": "ttyACM0"
+    },
+    "protocol": "serial",
+    "protocolLabel": "Serial Port (USB)"
+  }
+}
+```
+
+It basically provides the same information as the `list` event but for a single port. After calling `START_SYNC` an
+initial burst of add events must be generated in sequence to report all the ports available at the moment of the start.
+
+The `remove` event looks like the following:
+
+```JSON
+{
+  "eventType": "remove",
+  "port": {
+    "address": "/dev/ttyACM0",
+    "protocol": "serial"
+  }
+}
+```
+
+The content is straightforward, in this case only the `address` and `protocol` fields are reported.
+
+If the information about a port needs to be updated the discovery may send a new `add` message for the same port address
+and protocol without sending a `remove` first: this means that all the previous information about the port must be
+discarded and replaced with the new one.
+
+#### Invalid commands
+
+If the client sends an invalid or malformed command, the discovery should answer with:
+
+```JSON
+{
+  "eventType": "command_error",
+  "error": true,
+  "message": "Unknown command XXXX"
+}
+```
+
+### Board identification
+
+The `properties` associated to a port can be used to identify the board attached to that port. The algorithm is simple:
+
+- each board listed in the platform file `boards.txt` may declare a set of `upload_port.*` properties
+- if each `upload_port.*` property has a match in the `properties` set coming from the discovery then the board is a
+  “candidate” board attached to that port.
+
+Some port `properties` may not be precise enough to uniquely identify a board, in that case more boards may match the
+same set of `properties`, that’s why we called it “candidate”.
+
+Let’s see an example to clarify things a bit, let's suppose that we have the following `properties` coming from the
```

Comment:
```suggestion
  "candidate" board attached to that port.

Some port `properties` may not be precise enough to uniquely identify a board, in that case more boards may match the
same set of `properties`, that's why we called it "candidate".

Let's see an example to clarify things a bit, let's suppose that we have the following `properties` coming from the
```
I don't trust this highfalutin "smart quotes" stuff and we don't use it in the rest of the documentation. It was introduced from Discourse, which forces them on you.